### PR TITLE
Sequence length + Md5sum fields

### LINKED ISSUE #1509: Tripal 4 update YAML for various missing fields
# BUG/ERROR report

### Issue description

Currently publication content type has no fields defined except for "Authored by".
Presumably these are missing from
`tripal_chado/config/install/tripal.tripalfield_collection.default_chado.yml`

Other fields are also missing from various content types

#### Steps to reproduce
`/bio_data/add/bio_data_6`
 
#### Error messages and screenshots
add-publication

### PR DESCRIPTION
**Pull request by pdtouch**
#

Issue #1509

## Description
Modified the tripal_chado tripal_collections YAML file to include new fields into the gene page bundle.

## Testing?
- To test the YAML you need to build a new docker on this branch
```
n=1509
git clone https://github.com/tripal/tripal tripal-$n
git checkout tv4g1-1509-sequence-fields
docker build --tag=tripaldocker:testing-$n --build-arg drupalversion="10.1.x-dev" --file tripaldocker/Dockerfile-php8.2-pgsql13 ./
docker run --publish=8082:80 -tid --name=${n} --volume=/home/admin/git/tripal-$n:/var/www/drupal9/web/modules/contrib/tripal --volume=/home/admin/docker_prepare:/usr/local/bin/docker_common tripaldocker:testing-$n
```
- Log in to http://localhost:8082
- Create an organism record
- Create a gene record and include a sequence e.g. `actg`
- The sequence should appear along with sequence length and sequence MD5 checksum

### CODE REVIEW

File: tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    foreach($items as $delta => $item) {
+      $content = "<dl class=\"tripal-dl\">";
+<<<<<<< HEAD
+=======
+      $prog_vers_val = $item->get( 'programversion' )->getString() ;
+      if ( !empty( $prog_vers_val ) ) {
+        $content .= "<dt>Software version</dt><dd>: " . $prog_vers_val . " </dd>";
+      }
+>>>>>>> b73440286faed0b3cf68dec7fdff3e2bb4528450
```
Comment (dsenalik) [inline tripal_chado/src/Plugin/Field/FieldFormatter/ChadoLocalSourceDataFormatterDefault.php]: You need to remove the three merge conflict lines <<<< ===== and >>>>>

File: tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      'chado_column' => 'analysis_id',
+    ]);
+<<<<<<< HEAD
+=======
+    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'programversion', $prog_vers_term, 100, [
+      'action' => 'store',
+      'chado_table' => 'analysis',
+      'chado_column' => 'programversion',
+    ]);
+>>>>>>> b73440286faed0b3cf68dec7fdff3e2bb4528450
```
Comment (dsenalik) [inline tripal_chado/src/Plugin/Field/FieldType/ChadoLocalSourceDataDefault.php]: Remove <<<<< ===== >>>>> lines

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    ];
+<<<<<<< HEAD
+=======
+
+    $elements['programversion'] =  [
+      '#title' => "Software version",
+      '#type' => 'textfield',
+      '#default_value' => $item_vals['programversion'] ?? '',
+    ];
+>>>>>>> b73440286faed0b3cf68dec7fdff3e2bb4528450
```
Comment (dsenalik) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoLocalSourceDataWidgetDefault.php]: same here

**Review by dsenalik — DISMISSED** [review]
Code changes requested. I will build a docker for functional testing...

**Comment by dsenalik** [comment]
I added a testing procedure to this PR, but unfortunately the length and checksums are not working
gene-not-working

**Comment by dsenalik** [comment]
Debugging this, it seems that the checksum is generated, but is not found in the chado.feature table.
I suspect that maybe the checksum field is interfering and saving its null, somehow overriding saving the md5checksum from the sequence residues field. Maybe the order is important and that is why it is working when these three fields are added manually.
not-saving
Order of the fields from the YAML default installation:
wrong-order-maybe
When the docker is built, they appear to be added in the "correct" order
```
#27 158.2  [notice] Added field, "gene_sequence", to content type: "gene".
#27 158.7  [notice] Added field, "gene_length", to content type: "gene".
#27 159.3  [notice] Added field, "gene_sequence_md5_checksum", to content type: "gene".
```


Okay, the problem is that the length and checksum fields have "functional" widgets and all three fields are fighting over who gets to save last. The solution will be to disable those two widgets, they should not do anything, these are "read-only" fields.

**Comment by laceysanderson** [comment]
Interesting! I was actually noticing in making documentation that we didn't have a way to indicate that a field should read a column in the base table but not write. Currently we only have store as the join clearly doesn't work and is not meant for base table columns.

If making the widget non-functional (i.e. remove the form elements) doesn't work then this may need a small update to chado storage for this case.

**Comment by dsenalik** [comment]
If the widgets are disabled, then the main field properly updates the sequence, length, and checksum and it is stored in the feature table.
However, the length and checksum fields do not appear. I am sure now that this is the problem of fields not picking up records not added through the field, as you @laceysanderson mentioned in this comment. https://github.com/tripal/tripal/issues/1546#issuecomment-1642429013
Until we get that fixed, I think this PR will have to stay on hold.
I think the field "accidentally" worked before because by chance the order was right, and the other widgets caused the appropriate records to get generated in the `tripal_entity__gene_sequence_md5_checksum` and `tripal_entity__gene_length` tables, but the main field by chance saved last and generated the correct data.

An update, I am no longer 100% sure this is the reason, so further investigation is required...

**Comment by laceysanderson** [comment]
To use the new functionality in PR #1663, you will need to make the following changes:

```diff
diff --git a/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceChecksumDefault.php b/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceChecksumDefault.php
index de2b082fb..7b4896beb 100644
--- a/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceChecksumDefault.php
+++ b/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceChecksumDefault.php
@@ -79,12 +79,12 @@ class ChadoSequenceChecksumDefault extends ChadoFieldItemBase {
         'chado_column' => 'feature_id'
     ]);
     $properties[] =  new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'seqlen', $seqlen_term, [
-      'action' => 'store',
+      'action' => 'read_value',
       'chado_column' => 'seqlen',
       'chado_table' => 'feature'
     ]);
     $properties[] =  new ChadoBpCharStoragePropertyType($entity_type_id, self::$id, 'md5checksum', $md5checksum_term, $md5_checksum_len, [
-      'action' => 'store',
+      'action' => 'read_value',
       'chado_column' => 'md5checksum',
       'chado_table' => 'feature'
     ]);
diff --git a/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceLengthDefault.php b/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceLengthDefault.php
index 3463716b5..cadfcad80 100644
--- a/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceLengthDefault.php
+++ b/tripal_chado/src/Plugin/Field/FieldType/ChadoSequenceLengthDefault.php
@@ -16,7 +16,7 @@ use Drupal\tripal_chado\TripalStorage\ChadoIntStoragePropertyType;
  *   default_formatter = "chado_sequence_length_formatter_default"
  * )
  */
- 
+
 
 class ChadoSequenceLengthDefault extends ChadoFieldItemBase {
 
@@ -73,7 +73,7 @@ class ChadoSequenceLengthDefault extends ChadoFieldItemBase {
         'chado_column' => 'feature_id'
     ]);
     $properties[] =  new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'seqlen', $seqlen_term, [
-      'action' => 'store',
+      'action' => 'read_value',
       'chado_column' => 'seqlen',
       'chado_table' => 'feature'
     ]);
```

This PR also needs these fields to be added to the other feature-based content types. Specifically, mRNA, genetic marker, sequence variant, heritable phenotypic marker, QTL using the YAML in the same way you did for gene.

**Comment by pdtouch** [comment]
Requested changes are incorporated and testing completed successfully.
The Sequence Length and Checksum fields were added to other tripal content types :
`mRNA, genetic marker, sequence variant, heritable phenotypic marker, QTL` in addition to `Gene` 

Screenshot 2023-10-17 at 7 19 20 PM

Screenshot 2023-10-17 at 7 15 25 PM

Screenshot 2023-10-17 at 7 15 38 PM

**Review by laceysanderson — APPROVED** [review]
✅ Code looks good
✅ Requested changes were addressed
✅ All 3 fields showed up by default on all feature-based content types (i.e. gene, mRNA, genetic marker, sequence variant, QTL, heritable phenotypic marker)
✅ Fields correctly saved sequence residues, calculated length and mD5sum (see screenshot)
✅ No errors when creating content without the sequence set. sequence length set to 0 and md5sum hidden.
✅ No errors when content edited and sequence removed. sequence length updated to 0 and md5sum hidden.

I think this is ready to merge!

Editing to add the screenshot I mentioned 🙈 
Screenshot 2023-11-01 at 12 51 20 PM

**Review by dsenalik — APPROVED** [review]
I did a quick test too and it worked for me

**Pull request merged**